### ISSUE TITLE
FSDP Blocks LoRA Gradient Flow in Layers 1-31 (1% Coverage)

### ISSUE THREAD

**Issue by nelish01** [meta-pytorch/torchforge#592]
# FSDP Breaks Gradient Flow to LoRA Layers 1-31

## Issue Summary

**Torchforge's FSDP integration completely blocks gradient flow to LoRA parameters in layers 1-31, causing only layer 0 to receive gradients during backpropagation.**

### Impact
- **Standalone LoRA (no FSDP)**: 100% gradient coverage (32/32 layers)
- **Torchforge with FSDP**: 3.1% gradient coverage (1/32 layers)
- **Result**: 96.9% of LoRA parameters remain frozen during training

---

### Test 1: Standalone FP32 LoRA (Control Group) - ✅ WORKS

**Environment**: Pure PyTorch + HuggingFace Transformers (NO Torchforge, NO FSDP)

**Script**: [`examples/lora_training/minimal_eager_fp32_test.py`](examples/lora_training/minimal_eager_fp32_test.py)

**Configuration**:
- Model: Llama 3.1 8B
- LoRA rank: 64, alpha: 128
- Target: All 32 layers × 3 MLP projections (w1, w2, w3)
- Expected LoRA parameters: 192 (32 layers × 3 modules × 2 weights)
- Training mode: Eager FP32 (no AMP, no compile, no FSDP)

**Results**:
```
**SUCCESS! ALL LAYERS RECEIVE GRADIENTS!**

📊 Summary: Total layers: 32
Layers WITH gradients: 32
Layers WITHOUT gradients: 0
Coverage: 32/32 (100.0%)

Layer-by-layer gradient norms:
✅ Layer  0: avg grad = 3.796e-01
✅ Layer  1: avg grad = 1.488e+00
✅ Layer  2: avg grad = 3.725e-01
✅ Layer  3: avg grad = 3.928e-01
...
✅ Layer 30: avg grad = 3.968e-01
✅ Layer 31: avg grad = 9.030e-01
```

**Log file**: [`examples/lora_training/minimal_eager_fp32_test_results.log`](examples/lora_training/minimal_eager_fp32_test_results.log)

**Conclusion**: ✅ **LoRA implementation is correct** - all layers receive gradients in standalone mode.

---

### Test 2: Torchforge with FSDP - ❌ FAILS

**Environment**: Torchforge + FSDP

**Script**: [`apps/sft_lora/main_with_lora.py`](apps/sft_lora/main_with_lora.py)

**Configuration**:
- Model: Llama 3.1 8B
- LoRA rank: 16, alpha: 32
- Target: All 32 layers × 3 MLP projections (w1, w2, w3)
- Expected LoRA parameters: 192 (32 layers × 3 modules × 2 weights)
- Training mode: FSDP with mixed precision

**Results**:
```
❌ GRADIENT FLOW FAILURE

📊 GRADIENT FLOW SUMMARY:
  Total LoRA parameters:        192
  Params with gradients (>1e-8): 2 (1.0%)  ← ONLY 2 PARAMS!
  Params without gradients:      190 (99.0%)  ← 190 FROZEN!

📈 LAYER-BY-LAYER GRADIENT NORMS:
  Layer  0: 2 params, avg grad norm = 3.4192e+08  ← ONLY LAYER 0!
    - lora_a_weight: 6.0442e-03
    - lora_b_weight: 6.8384e+08

⚠️ PARTIAL gradient flow: 2/192
   → Problem is GRADIENT FLOW is BROKEN
   → Only 1/32 layers receiving gradients
   → Layers 1-31: ZERO gradients
```

**Log file**: [`apps/sft_lora/logs/training_20251118_123030.log`](apps/sft_lora/logs/training_20251118_123030.log)

**Loss behavior**:
```
Training Run 123030 (with FSDP):
  Initial Loss (Step 0):  12.3135
  Final Loss (Step 99):   3.2319
  Total Loss Drop:        9.0816
  Percent Reduction:      73.75%

Note: Despite 73% loss reduction, this is achieved with ONLY layer 0 learning.
With proper gradient flow to all 192 parameters, we would expect 85-90% reduction.
```

**Conclusion**: ❌ **FSDP breaks gradient flow** - only layer 0 receives gradients.

---

## 📊 Side-by-Side Comparison

| Metric | Standalone (No FSDP) | Torchforge + FSDP | Difference |
|--------|---------------------|-------------------|------------|
| **Total LoRA Parameters** | 192 | 192 | Same |
| **Params with Gradients** | 192 (100%) | 2 (1.0%) | **-99%** ❌ |
| **Layers Learning** | 32/32 (100%) | 1/32 (3.1%) | **-96.9%** ❌ |
| **Layer 0 Gradient** | ✅ Yes | ✅ Yes | Same |
| **Layers 1-31 Gradients** | ✅ Yes | ❌ **ZERO** | **BROKEN** |
| **Loss Reduction** | N/A | 73.75% | Limited by frozen layers |

---

### What We Verified

1. ✅ **LoRA parameters exist**: All 192 parameters are created correctly
2. ✅ **`requires_grad=True` is set**: Verified for all LoRA parameters before FSDP
3. ✅ **Optimizer sees parameters**: All 192 LoRA params registered with optimizer
4. ✅ **Forward pass works**: Model processes inputs correctly through all layers
5. ❌ **Backward pass fails**: Gradients don't flow past layer 0

### The Problem

**FSDP's gradient computation doesn't properly propagate through LoRA modules in layers 1-31.**

This is NOT:
- ❌ A `requires_grad` configuration issue (we verified all params have `requires_grad=True`)
- ❌ An optimizer registration issue (all params are in optimizer)
- ❌ A LoRA implementation bug (standalone test proves LoRA works perfectly)

So it seems a framework-level FSDP integration bug in how Torchforge handles gradient computation through LoRA-wrapped modules.

---

### Fix 1: Comment out `set_trainable_params()`
**Rationale**: Remove potential parameter freezing
**Result**: ❌ Still only 2/192 params receive gradients

### Fix 2: Pattern-based `requires_grad` setting
**Rationale**: Avoid name-matching issues after FSDP wrapping
**Result**: ❌ Still only 2/192 params receive gradients

### Fix 3: Monkey-patch to apply LoRA before FSDP
**Rationale**: Ensure LoRA is applied before FSDP wrapping
**Result**: ❌ Still only 2/192 params receive gradients

**Conclusion**: The issue requires framework-level changes to Torchforge's FSDP backward pass implementation.

---

### Standalone Test (Proves LoRA Works)
```bash
# Run standalone FP32 test
cd /path/to/torchforge
conda run -n your_env python examples/lora_training/minimal_eager_fp32_test.py

# Expected output: "✅✅✅ SUCCESS! ALL LAYERS RECEIVE GRADIENTS!"
```

**File**: `examples/lora_training/minimal_eager_fp32_test.py`

### Torchforge FSDP Test (Shows Bug)
```bash
# Run Torchforge LoRA training
cd /path/to/torchforge/apps/sft_lora
bash train.sh

# Check gradient flow in logs
grep -A20 "GRADIENT FLOW SUMMARY" logs/training_*.log
```

**Files**:
- Training script: `apps/sft_lora/main_with_lora.py`
- Config: `apps/sft_lora/llama3_8b_all_layers_production.yaml`
- Launch script: `apps/sft_lora/train.sh`

---

### Expected Behavior
When LoRA is applied to all 32 layers:
- ✅ All 192 LoRA parameters should receive gradients during backward pass
- ✅ Loss should decrease significantly (85-90% reduction over 100 steps)
- ✅ All layers should learn and contribute to the fine-tuning

### Actual Behavior
- ❌ Only 2/192 LoRA parameters receive gradients (layer 0 only)
- ❌ Loss reduction limited to 73% (constrained by single learning layer)
- ❌ Layers 1-31 completely frozen (0.00e+00 gradient norm)

---

## 💻 Environment Details

**Torchforge Version**: `main` branch (as of Nov 2024)
**PyTorch Version**: 2.x (with FSDP support)
**Model**: Llama 3.1 8B
**Hardware**: NVIDIA GPU with CUDA
**Python**: 3.10+

---

### Gradient Flow Logging (Added to Code)

We added comprehensive gradient flow diagnostics to track this issue:

```python
def _log_gradient_norms(self):
    """Log gradient norms for all LoRA parameters."""
    layer_grad_stats = {}
    total_lora_params = 0
    params_with_grads = 0

    for model_part in self.model_parts:
        for name, param in model_part.named_parameters():
            if "lora_" in name:
                total_lora_params += 1
                if param.grad is not None:
                    grad_norm = param.grad.norm().item()
                    if grad_norm > 1e-8:
                        params_with_grads += 1
                        # Track per-layer stats...
```

**Output shows**:
```
Layer  0: ✅ Gradients present (norm = 3.4e+08)
Layer  1: ❌ grad_norm = 0.00e+00 (NEAR ZERO!)
Layer  2: ❌ grad_norm = 0.00e+00 (NEAR ZERO!)
...
Layer 31: ❌ grad_norm = 0.00e+00 (NEAR ZERO!)
```

---

Based on our analysis, the really wanted to understand that with Torchforge as per the below

1. **FSDP backward pass hooks**: How gradients propagate through FSDP-wrapped LoRA modules
2. **Parameter sharding**: Whether LoRA parameters are being incorrectly sharded across ranks
3. **Gradient accumulation**: If FSDP's gradient accumulation skips LoRA parameters in layers 1-31
4. **Module wrapping order**: Whether FSDP wrapping interferes with LoRA's forward/backward hooks

---

## 📎 Evidence Files

**All evidence files are available in this GitHub Gist:**  
https://gist.github.com/nelish01/f5ed4cd29c5caac91e5ce1284c2285bf

### Files Included:
1. ✅ **`minimal_eager_fp32_test_results.log`** - Standalone test (100% gradient coverage)
2. ❌ **`training_fsdp_failure_excerpt.log`** - Torchforge FSDP test (1% gradient coverage)
3. **`minimal_eager_fp32_test.py`** - Standalone test script
4. **`llama3_8b_all_layers_production.yaml`** - LoRA configuration
5. **`torchforge_lora_recipe_excerpt.py`** - Torchforge implementation excerpts

---

### Step 1: Run standalone test (expects 100% success)
```bash
cd torchforge
python examples/lora_training/minimal_eager_fp32_test.py
```

**Expected output:**
```
✅✅✅ SUCCESS! ALL LAYERS RECEIVE GRADIENTS! ✅✅✅
Coverage: 32/32 (100.0%)
```

### Step 2: Run Torchforge FSDP test (shows bug)
```bash
cd torchforge/apps/sft_lora
bash train.sh

# Check gradient flow
grep -A20 "GRADIENT FLOW SUMMARY" logs/training_*.log
```

**Expected output:**
```
❌ GRADIENT FLOW FAILURE
Params with gradients (>1e-8): 2 (1.0%)
Only 1/32 layers receiving gradients
```

---

## 📞 Contact

**Reporter**: Nilesh  
**Date**: November 18, 2024  
**Environment**: Llama 3.1 8B, PyTorch 2.x, CUDA GPU
---

**Comment by casteryh** [comment]
Can't repro as repro code and logs are not provided after requesting offline. (Links attached don't work)
